Changed code structure, added tab autocompletion

TheodorJ opened the pull request.

TheodorJ committed:
implemented init_autolab_command_map

TheodorJ committed:
Added bool instructor_command field to command_info

TheodorJ committed:
Rewrote print_help

TheodorJ committed:
main now uses command maps

TheodorJ committed:
Added autocomplete script

TheodorJ committed:
Merge branch 'master' into autocomplete-imp

TheodorJ commented:
Just tested -h, everything still works

zyx-billy committed:
fix unrecognized command error message

zyx-billy commented on src/cmd/cmdmap.cpp:
```diff
@@ -1,42 +1,131 @@
+  std::vector<std::string> valid_commands = {"status", "download", "submit",
+                                             "courses", "assessments", "asmts",
+                                             "problems", "scores", "feedback",
+                                             "enroll"};
```
are there any additional uses for the valid_commands vector?
I'm thinking we may be able to omit this, and instead have print_help() loop through the values of the map directly. This will also avoid the need to check if a command exists each time we get its usage string.

TheodorJ replied:
Yeah, it's not necessary. The main "advantage" of the vector is to make sure we print the values in a specified order, but that's really a matter of preference.

zyx-billy replied:
I see. I guess it does make sense to be able to order them, but having another vector just seems a bit redundant... What do you think if we replaced the map with an array of command_info structs in order? Would that make it better or worse

TheodorJ replied:
Huh, that would make it more efficient, but then we don't have the intuition of retrieving command info from the command name. We'd also sacrifice the O(1) lookup time for O(n), but when our n is equal to 10... I guess getting rid of the ordering isn't too big of a loss, if we ditch the redundancy.

zyx-billy replied:
yeah I wouldn't worry about the performance with 10 elements since the map uses a red-black tree internally (log n), which also needs to be self-balanced during insertion. I do agree that the intuition of "mapping" command names to command functions is quite useful. I guess we can leave the map as is and remove the order vector. Iterating through the map is guaranteed to be in lexicographical order anyway, so it's not bad at all.

TheodorJ replied:
Oh, that's no problem then. I can take care of fixing this and removing the unnecessary comment below in the morning. I'll commit on top then.

zyx-billy replied:
sounds good, thanks!

zyx-billy commented on src/cmd/cmdmap.cpp:
```diff
@@ -1,42 +1,131 @@
-  // Why am I returning an int again?
-  return 0;
-}
+  // There's gotta be a better way to do this
```
did you mean to come up with a better way? :)

TheodorJ replied:
Oh, haha. Yeah, that was a complete mess back when I was trying to use malloc() to instantiate the map. I guess we can delete the comment now :P

zyx-billy replied:
haha alright

TheodorJ committed:
Got rid of unnecessary comment

TheodorJ committed:
Merge branch 'autocomplete-imp' of https://github.com/autolab/autolab-cli into autocomplete-imp

gotta merge thx

TheodorJ committed:
Removed command vector

zyx-billy committed:
remove redundant error checking

with the command vector removed, there's no more need for querying the map when looping through it. Also we can query the map with the command driectly to check if a command exists.

zyx-billy commented:
looks cool! The new alias map seems like a straightforward way of dealing with aliases.
Part of the redundancy I was referring to yesterday was checking if a valid command exists in the info_map when getting its usage string. Now that we loop through info_map directly, there's no need to query the map for the command_info any more since we get it by default with the iterator. I've updated the code with what I mean. (with some minor style changes to be more concise).

zyx-billy commented:
Before merging though, the seemingly random amount of spaces in the usage strings are kind of annoying... Now that the strings are all spread out, it's going to be quite hard to align them if we add additional commands in the future. I'm going to try and get them aligned...

TheodorJ commented:
Sounds like something we can automate -- the length of the string of spaces is equal to the longest command plus one character

zyx-billy commented:
ah that's true. I guess we could do that. I was originally going to propose a dead simple list initialization
![image](https://user-images.githubusercontent.com/3676913/38121590-344243da-339e-11e8-83b6-dfa2c345d605.png)
But I guess doing it dynamically inside print_help() works too (that's actually how the help text for the individual commands are generated rn). The only problem is we have to somehow reverse the alias_map.

zyx-billy committed:
align command usages

also bump version number
